### ISSUE TITLE
Don't store addons in top-level of addons reducer

### ISSUE THREAD

**Issue by tofumatt** [mozilla/addons-frontend#3421]
Right now we store addons in the top of the addons reducer rather than organzing them in a dictionary under a key of the main reducer state object. Right now we have:

```js
{
  addonsReducer: {
    addonSlug: { [...addon] },
    anotherAddonSlug: { [...addon] },
  },
  [...]
}
```

But this prevents us from storing any other info in this reducer, like languagePack slugs as we wanted to in https://github.com/mozilla/addons-frontend/pull/3289#discussion_r143288040

We want:

```js
{
  addonsReducer: {
    bySlug: {
      addonSlug: { [...addon] },
      anotherAddonSlug: { [...addon] },
    },
    otherReducerData,
  },
  [...]
}
```

In addition and since #3657, we store an `addon` twice in the state. We should not do this. Instead, we should store the `addon` only once and reference it by ID in all other parts of the state (a `slug` could change).

**Comment by willdurand** [comment]
> In addition and since #3657, we store an addon twice in the state. We should not do this. Instead, we should store the addon only once and reference it by ID in all other parts of the state (a slug could change).

Yeah, this is causing:

1. twice the amount of data supplied to the `LanguageTools` component and many React warnings because keys are duplicate
2. the languages packs to be indexed twice with numeric values: one being the `id` and the other being the index (natural integer) because it uses `loadAddonResults()`, which does not rely on normalized (normalizr) data

**Comment by kumar303** [comment]
Since this issue doesn't cover it fully, here is my wish list for the implementation :)

- we should only store the add-on object once, by ID
- all other mappings should link to the add-on. For example, `bySlug` should be a mapping like `{ slugOne: 1234, slugTwo: 5432 }` where the key is the add-on slug and the value is the ID of the add-on.

**Comment by willdurand** [comment]
I have removed the `qa: not needed` badge because the patch is sensitive. For QA, please make sure:

- Disco Pane continues to work as expected
- AMO continues to work as expected (add-ons load fine, add-ons can be installed)

**Comment by vcarciu** [comment]
@willdurand : not sure if this is related to this bug , but searches in disco pane are not working as expected for me in -dev.  With all disco pane prefs set to -dev, if you search for "firefox" for example , there are no results, even if we have many compatible webextensions found with same search term on AMO. Please let me know if you think that it is related to this issue(you can see in video that some searchterms are working but another one not) : 
![discosearch](https://user-images.githubusercontent.com/1583842/33218765-57ec1f54-d147-11e7-8809-08ba459e22d9.gif)

**Comment by willdurand** [comment]
mmh I dont think this is part of disco pane but might be wrong..

**Comment by willdurand** [comment]
@vcarciu I confirm it is not related.

**Comment by vcarciu** [comment]
@willdurand : thanks, I will mark this as verified then and I will open a separate issue.
